Reimplemented MonadError[FreeT[...]] to be correct

djspiewak opened the pull request:
There's an explanation of this in the comments in the code. I retained the old implementation for binary compatibility, but it is somewhat unfortunate since any code which is hitting that is outright-broken and people may not know it. Then again, I'm not sure how many people (other than me) are using `FreeT`, so maybe it's okay either way.

Fixes #3240

LukaJCB commented on free/src/main/scala/cats/free/FreeT.scala:
```diff
@@ -220,14 +220,67 @@ object FreeT extends FreeTInstances {
+      override def M = E
+
+      /*
+       * Quick explanation... The previous version of this function (retained about for
```
typo: `...retained **above** for...`

kubukoz commented:
Is it possible to adjust the generators / equalities used in tests to verify that the old implementation does indeed break a law?

kailuowang commented:
Thanks for spotting and fixing this bug. 
A couple questions if you don't mind: 
It seems that making it `private[this]` still breaks BC, probably need to stick to package private. 
Why keeping the old faulty implementation (v.s. letting the old method delegate to the new correct one)? 
It kind of itches that a faulty `handleErrorWith` implementation doesn't break the `ApplicativeErrorLaws`. There is no way to add a new law to capture this right?

djspiewak commented:
@kubukoz I've been wondering the same thing myself. I tried fiddling around with some sort of distributivity law for `handleErrorWith` but it's surprisingly subtle, because you still need to ensure ordered short-circuiting. I would definitely like to codify this better.

@kailuowang See above. :-) I want to add a law. Thoughts welcome. My attempt was something like this:

```scala
@kubukoz I've been wondering the same thing myself. Something like this perhaps in `MonadErrorLaws`? We can't work with `ApplicativeErrorLaws` unfortunately because we don't want to accidentally force `handleErrorWith` to somehow handle errors raised by its handler:

```scala
def handleErrorWithDistributesOverAp[A, E](fa1: F[A], fa2: F[A], h: E => F[A]): F[A] =
  F.handleErrorWith(fa1 *> fa2)(h) <-> (F.handleErrorWith(fa1)(h) *> F.handleErrorWith(fa2)(h))
```

That's not actually right, though, because `fa1` should short-circuit `fa2` if it errors, whereas it will not in the right side.

> Why keeping the old faulty implementation (v.s. letting the old method delegate to the new correct one)?

The constraints are different. The new requires a `Functor[S]`. I *am* tempted to just add a `sys.error` though, because any code using the old implementation is broken in a very dangerous way.

> It seems that making it private[this] still breaks BC, probably need to stick to package private.

Actually it has to be public. The problem is the static forwarder. I've fixed it now.

kailuowang commented:
Thanks @djspiewak I missed the new constraint. This is source breaking then, I am adding the label just so that it doesn't surprise users.

djspiewak commented:
How about this as a law?

```scala
def attemptDistributivity[A, B](fa: F[A], f: A => F[B]): F[Either[E, B]] =
  F.attempt(F.flatMap(fa)(f)) <-> F.flatMap(F.attempt(fa))(_.traverse(f.andThen(F.attempt)))
```

djspiewak commented:
(btw, please don't merge this just yet; even aside from the discussion about a new law, which is important, I have some ancillary evidence that this isn't appropriately stack-safe.

djspiewak commented:
@kailuowang @kubukoz Okay, I've added a pair of laws in ApplicativeError which capture this case. I've verified that they do indeed fail on `FreeT` in master. They aren't *quite* as strong as what I would like (explained in a comment), but I think they should appropriately constrain any non-pathological implementations.

I've also added a `Defer` for `FreeT`, which was previously absent, and used it to ensure the stack-safety of `handleErrorWith` even when the suspension monad does not provide a stack-safe `flatMap`. An obvious objection here is that, if the suspension monad is not stack-safe, then presumably stack-safety in `handleErrorWith` is just delaying the inevitable. However, it's entirely possible to run `handleErrorWith` on `FreeT` when the suspension monad is not stack-safe, and then subsequently `mapK` into a stack-safe context prior to interpretation. That use-case was impossible with my initial implementation.

kubukoz commented on laws/src/main/scala/cats/laws/ApplicativeErrorLaws.scala:
```diff
@@ -54,6 +54,21 @@ trait ApplicativeErrorLaws[F[_], E] extends ApplicativeLaws[F] {
+  /*
+   * These laws, taken together with applicativeErrorHandle, show that errors dominate in
+   * ap, *and* show that handle has lexical semantics over ap. F.unit is used in both laws
+   * because we don't have another way of expressing "an F[_] which does *not* contain any
```
Another way might be `Arbitrary[F[_]].map(_.attempt)`... But that's probably problematic in other ways (depending on the behavior of attempt)

djspiewak replied:
Well, the problem is then you need a function `E => A`. That's not doable without breaking binary compatibility.

kubukoz replied:
Why binary compatibility? I fail to see that here...

djspiewak replied:
In order to have an `Arbitrary[E => A]`, we need a `Cogen[E]` and an `Arbitrary[A]`. We have the former, but not the latter (in `ApplicativeErrorTests`). In order to *get* the latter, we would need to add it as a parameter to the `applicativeError` function, which breaks binary compatibility.

kubukoz replied:
Ohhh. Yeah, you're absolutely right.
